src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -751,6 +751,81 @@ public void PresentChain_CheckpointsEnabled_MarkToDownloadWhenTwoCheckpointsPres
+            List<BlockHeader> listOfCurrentChainHeaders =
+                testContext.ChainedHeaderToList(extendedChainTip, initialChainOfTenBlocks + extendChainByTwentyBlocks).Take(extendChainByTwentyBlocks).ToList();
+            ConnectNewHeadersResult connectNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfCurrentChainHeaders);
+            ChainedHeader chainedHeaderDownloadFrom = connectNewHeadersResult.DownloadFrom;
+            ChainedHeader chainedHeaderDownloadTo = connectNewHeadersResult.DownloadTo;
+
+            chainedHeaderDownloadFrom.HashBlock.Should().Be(initialChainTip.Next[0].HashBlock); // h11
+            chainedHeaderDownloadTo.HashBlock.Should().Be(listOfCurrentChainHeaders.Last().GetHash());
+
+            // Blocks before X-20 (h1->h10) are FV.
+            ChainedHeader chainedHeader = initialChainTip;
+            Assert.Equal(chainedHeader.Height, assumeValidBlockHeight - 20);
+            ValidationState expectedState = ValidationState.FullyValidated;
+            while (chainedHeader.Height > 0)
+            {
+                chainedHeader.BlockValidationState.Should().Be(expectedState);
+                chainedHeader = chainedHeader.Previous;
+            }
+
+            // Headers up to block X (h30) + some more (h31->h35) are presented.
+            const int extendChainByFiveBlocks = 5;
+            extendedChainTip = testContext.ExtendAChain(extendChainByFiveBlocks, extendedChainTip);
+            listOfCurrentChainHeaders =
+                testContext.ChainedHeaderToList(extendedChainTip, assumeValidBlockHeight + extendChainByFiveBlocks);
+            connectNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfCurrentChainHeaders);
+
+            // All from X-10 (h21->h30) are marked for download.
+            ChainedHeader fromHeader = chainedHeaderTree.GetChainedHeadersByHash().Values.First(x => x.Height > assumeValidBlockHeight - 10);
+            
+            chainedHeaderDownloadFrom = connectNewHeadersResult.DownloadFrom;
+            chainedHeaderDownloadTo = connectNewHeadersResult.DownloadTo;
+            
+            chainedHeaderDownloadFrom.HashBlock.Should().Be(fromHeader.HashBlock);
+            chainedHeaderDownloadTo.HashBlock.Should().Be(extendedChainTip.HashBlock);
+
+            // Make sure that all blocks before assumevalid block that are not FV or PV are marked assumevalid.
+            var allowableBlockValidationStates = new[]
+            {
+                ValidationState.PartiallyValidated, ValidationState.FullyValidated, ValidationState.AssumedValid
+            };
+
+            IEnumerable<ChainedHeader> headersBeforeAssumeValid = 
+                chainedHeaderTree.GetChainedHeadersByHash().Values.Where(x => x.Height < assumeValidBlockHeight);
+
+            Assert.True(headersBeforeAssumeValid.All(x =>
+                allowableBlockValidationStates.Contains(x.BlockValidationState)));
```

Comment:
This is little bit funny, as you took the description literally, which actually is not that bad, but it would be more sensible to implement this in different way. I'm only talking about this last part from line 816. He would be actually wanted to do, is just Traverse the chain from the assumed valid block and check that every block, including the assumed valid block, is marked as assumed valid, down to the block 11 (included).

Reply:
Done